hip/components/cooperative_groups.hip.hpp

```diff
@@ -0,0 +1,442 @@
+    } data_;
+};
+
+
+}  // namespace detail
+
+
+namespace detail {
+
+
+// Adds generalized shuffles that support any type to the group.
+template <typename Group>
+class enable_extended_shuffle : public Group {
+public:
+    using Group::Group;
+    using Group::shfl;
+    using Group::shfl_down;
+    using Group::shfl_up;
+    using Group::shfl_xor;
+
+#define GKO_ENABLE_SHUFFLE_OPERATION(_name, SelectorType)                   \
+    template <typename ValueType>                                           \
+    __device__ __forceinline__ ValueType _name(const ValueType &var,        \
+                                               SelectorType selector) const \
+    {                                                                       \
+        return shuffle_impl(                                                \
+            [this](uint32 v, SelectorType s) {                              \
+                return static_cast<const Group *>(this)->_name(v, s);       \
+            },                                                              \
+            var, selector);                                                 \
+    }
+
+    GKO_ENABLE_SHUFFLE_OPERATION(shfl, int32)
+    GKO_ENABLE_SHUFFLE_OPERATION(shfl_up, uint32)
+    GKO_ENABLE_SHUFFLE_OPERATION(shfl_down, uint32)
+    GKO_ENABLE_SHUFFLE_OPERATION(shfl_xor, int32)
+
+#undef GKO_ENABLE_SHUFFLE_OPERATION
+
+private:
+    template <typename ShuffleOperator, typename ValueType,
+              typename SelectorType>
+    static __device__ __forceinline__ ValueType
+    shuffle_impl(ShuffleOperator intrinsic_shuffle, const ValueType &var,
+                 SelectorType selector)
+    {
+        static_assert(sizeof(ValueType) % sizeof(uint32) == 0,
+                      "Unable to shuffle sizes which are not 4-byte multiples");
+        constexpr auto value_size = sizeof(ValueType) / sizeof(uint32);
+        ValueType result;
+        auto var_array = reinterpret_cast<const uint32 *>(&var);
+        auto result_array = reinterpret_cast<uint32 *>(&result);
+#pragma unroll
+        for (std::size_t i = 0; i < value_size; ++i) {
+            result_array[i] = intrinsic_shuffle(var_array[i], selector);
+        }
+        return result;
+    }
+};
+
+
+}  // namespace detail
+
+
+// Implementing this as a using directive messes up with SFINAE for some reason,
+// probably a bug in NVCC. If it is a complete type, everything works fine.
+template <size_type Size>
+struct thread_block_tile
+    : detail::enable_extended_shuffle<detail::thread_block_tile<Size>> {
+    using detail::enable_extended_shuffle<
+        detail::thread_block_tile<Size>>::enable_extended_shuffle;
+};
+
+
+// Only support tile_partition with 1, 2, 4, 8, 16, 32, 64 (hip).
+template <size_type Size, typename Group>
+__device__ __forceinline__ gko::xstd::enable_if_t<
+    (Size <= kernels::hip::hip_config::warp_size) &&
+        (kernels::hip::hip_config::warp_size % Size == 0),
+    thread_block_tile<Size>>
```

Comment:
The thread_block_tile is used as subwarp group, so it can only support (1, 2, 4, ..., warp_size).

Reply:
Sorry,  I do not understand. Where does this come from? In CUDA we do not have this limitation for example. Afaik, for HIP also shuffle etc can work with any size, no?

See the talk on lane here: https://github.com/ROCm-Developer-Tools/HIP/blob/master/docs/markdown/hip_kernel_language.md#warp-cross-lane-functions

Reply:
- For Hip, I do not find any documentation about it.
According to https://github.com/ROCm-Developer-Tools/HIP/blob/56193a7828a75cb5fc6ea9027c27648c317b9122/include/hip/hcc_detail/device_functions.h#L291, if width is not power of 2, the result is not expected. width - 1 will be the line mask, so it needs to be power of 2.
For example, `width = 3, shfl(var, 1)`. At line 1, `src + id & ~(width-1) = 1 + 1 & ~2 = 1 + 1 = 2` not `3`. Thus, the width should also be power of 2 and <= max_warp_size.
I use `Size <= config::warp_size && config::warp_size % Size == 0` to fit this rule.
I forget to add `Size > 0`. I will add it in next commit.


- For cuda, warpsize can only be the power of 2 and less than or equal to 32. (ref. https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#warp-notes)
Also, tiled_partition<Size> Size also has same limitation to use warp function.
line 807-824 of `cuda/cooperative_group.h`
```C
/**
 * class thread_block_tile<unsigned int Size>
 *
 * Statically-sized group type, representing one tile of a thread block.
 * The only specializations currently supported are those with native
 * hardware support (1/2/4/8/16/32)
 *
 * This group exposes warp-synchronous builtins.
 * Constructed via tiled_partition<Size>(class thread_block);
 */
template <unsigned int Size>
class thread_block_tile;
template <> class thread_block_tile<32> : public __thread_block_tile_base<32> { };
template <> class thread_block_tile<16> : public __thread_block_tile_base<16> { };
template <> class thread_block_tile<8>  : public __thread_block_tile_base<8> { };
template <> class thread_block_tile<4>  : public __thread_block_tile_base<4> { };
template <> class thread_block_tile<2>  : public __thread_block_tile_base<2> { };
template <> class thread_block_tile<1>  : public __thread_block_tile_base<1> { };
```

Reply:
@trost57 maybe you know where to find documentation about the possible shuffle commands?

Reply:
Thanks for the info @yhmtsai. I should have looked at the code implementation for HIP indeed. I guess we should also put a note somewhere in the cuda cooperative groups? I was confused about this point since jacobi generates kernels for all block sizes, but now I realize it passes in a power of two when using tiled partition.